app/server/erlang/pi_server_cue.erl

```diff
@@ -29,99 +34,137 @@
+                    {ok, NewInSocket} = gen_udp:open(InPort,
+                                                     [binary, {ip, loopback}]),
+                    ?MODULE:loop(State#{internal := true,
+                                        in_socket := NewInSocket})
             end;
 
         {internal, false} ->
-            case Internal of
-                true ->
+            case State of
+                #{internal := true,
+                  in_socket := InSocket,
+                  in_port := InPort} ->
                     log("Switching cue listener to open network~n"),
                     gen_udp:close(InSocket),
                     {ok, NewInSocket} = gen_udp:open(InPort, [binary]),
-                    ?MODULE:loop(NewInSocket, InPort, CueHost, CuePort, false, Enabled);
-                _ ->
-                    ?MODULE:loop(InSocket, InPort, CueHost, CuePort, false, Enabled)
+                    ?MODULE:loop(State#{internal := false,
+                                        in_socket := NewInSocket});
+                #{internal := false} ->
+                    ?MODULE:loop(State#{internal := false})
             end;
 
         {enabled, true} ->
             log("Enabling cue forwarding ~n"),
-            ?MODULE:loop(InSocket, InPort, CueHost, CuePort, Internal, true);
+            ?MODULE:loop(State#{enabled := true});
 
         {enabled, false} ->
             log("Disabling cue forwarding ~n"),
-            ?MODULE:loop(InSocket, InPort, CueHost, CuePort, Internal, false);
+            ?MODULE:loop(State#{enabled := false});
 
         {timeout, Timer, {forward, Time, Data, Tracker}} ->
-            send_forward(InSocket, Time, Data),
+            send_forward(maps:get(in_socket, State), Time, Data),
             pi_server_tracker:forget(Timer, Tracker),
-            ?MODULE:loop(InSocket, InPort, CueHost, CuePort, Internal, Enabled);
+            ?MODULE:loop(State);
 
         {forward, Time, Data} ->
-            send_forward(InSocket, Time, Data),
-            ?MODULE:loop(InSocket, InPort, CueHost, CuePort, Internal, Enabled);
+            send_forward(maps:get(in_socket, State), Time, Data),
+            ?MODULE:loop(State);
 
         {udp_error, _Port, econnreset} ->
             %% Should not happen, but can happen anyway on Windows
             debug(2, "got UDP ECONNRESET - ignored~n", []),
-            ?MODULE:loop(InSocket, InPort, CueHost, CuePort, Internal, Enabled);
+            ?MODULE:loop(State);
 
+        {system, From, Request} ->
+            %% handling system messages (like a gen_server does)
+            sys:handle_system_msg(Request, From,
+                                  maps:get(parent, State),
+                                  ?MODULE, [], State);
         Any ->
 	    log("Cue Server got unexpected message: ~p~n", [Any]),
-	    ?MODULE:loop(InSocket, InPort, CueHost, CuePort, Internal, Enabled)
+	    ?MODULE:loop(State)
 
     after ?IDLE_TIME ->
```

Comment:
Are these timeout clauses really needed?

Reply:
I could remove them in another PR later. Joe had them in there, I guess it was nice during development to see that the processes were alive even when they were not receiving any events.